--- SDI12.cpp	2025-01-25 18:19:51.651337996 +0100
+++ SDI12.cpp	2025-01-25 18:19:36.050886233 +0100
@@ -356,6 +356,7 @@
       {
         pinMode(_dataPin, INPUT);     // Turn off the pull-up resistor
         pinMode(_dataPin, OUTPUT);    // Pin mode = output
+        digitalWrite(MUX_PIN, LOW);
         digitalWrite(_dataPin, LOW);  // Pin state = low - marking
         setPinInterrupts(false);      // Interrupts disabled on data pin
         break;
@@ -364,6 +365,7 @@
       {
         pinMode(_dataPin, INPUT);   // Turn off the pull-up resistor
         pinMode(_dataPin, OUTPUT);  // Pin mode = output
+        digitalWrite(MUX_PIN, LOW);
         setPinInterrupts(false);    // Interrupts disabled on data pin
 #ifdef SDI12_CHECK_PARITY
         _parityFailure = false;  // reset the parity failure flag
@@ -372,6 +374,7 @@
       }
     case SDI12_LISTENING:
       {
+        digitalWrite(MUX_PIN, HIGH);
         digitalWrite(_dataPin, LOW);  // Pin state = low (turns off pull-up)
         pinMode(_dataPin, INPUT);     // Pin mode = input, pull-up resistor off
         interrupts();                 // Enable general interrupts
@@ -381,6 +384,7 @@
       }
     default:  // SDI12_DISABLED or SDI12_ENABLED
       {
+        digitalWrite(MUX_PIN, HIGH);
         digitalWrite(_dataPin, LOW);  // Pin state = low (turns off pull-up)
         pinMode(_dataPin, INPUT);     // Pin mode = input, pull-up resistor off
         setPinInterrupts(false);      // Interrupts disabled on data pin
